ui/org.eclipse.pde.junit.runtime/src/org/eclipse/pde/internal/junit/runtime/RemotePluginTestRunner.java

```diff
@@ -99,19 +102,34 @@ private static ClassLoader createJUnit5PluginClassLoader(String testPluginName)
 
 	private static List<Bundle> findTestEngineBundles() {
 		BundleContext bundleContext = FrameworkUtil.getBundle(RemotePluginTestRunner.class).getBundleContext();
-		List<Bundle> engineBundles = new ArrayList<>();
-		for (Bundle bundle : bundleContext.getBundles()) {
-			try {
-				BundleWiring bundleWiring = bundle.adapt(BundleWiring.class);
-				Collection<String> listResources = bundleWiring.listResources("META-INF/services", "org.junit.platform.engine.TestEngine", BundleWiring.LISTRESOURCES_LOCAL); //$NON-NLS-1$//$NON-NLS-2$
-				if (!listResources.isEmpty()) {
-					engineBundles.add(bundle);
-				}
-			} catch (Exception e) {
-				// check the next bundle
+		return Arrays.stream(bundleContext.getBundles()).filter(RemotePluginTestRunner::providesCompatibleTestEngine).collect(toCollection(ArrayList::new));
+	}
+
+	/**
+	 * Checks whether the bundle provides test engines.
+	 * Ensures that test engines that can be loaded from the bundle
+	 * are compatible with the TestEngine version in current scope.
+	 * Otherwise, the JUnit platform's call to the ServiceLoader for
+	 * retrieving available engines will fail.
+	 * Incompatibilities can happen, e.g., in Tycho builds, where
+	 * the org.eclipse.tycho.surefire.osgibooter bundle is found
+	 * that may provide a different JUnit platform version than the
+	 * one available via the Eclipse target platform.
+	 */
+	private static boolean providesCompatibleTestEngine(Bundle bundle) {
+		try {
+			BundleWiring bundleWiring = bundle.adapt(BundleWiring.class);
+			String testEngineClass = "org.junit.platform.engine.TestEngine"; //$NON-NLS-1$
+			Collection<String> engineProviders = bundleWiring.listResources("META-INF/services", testEngineClass, BundleWiring.LISTRESOURCES_LOCAL); //$NON-NLS-1$
+			if (!engineProviders.isEmpty()) {
+				Class<?> thisTestEngine = Class.forName(testEngineClass);
+				Class<?> bundleTestEngine = bundle.loadClass(testEngineClass);
+				return thisTestEngine != null && bundleTestEngine != null && thisTestEngine.isAssignableFrom(bundleTestEngine);
```

Comment:
It looks like `Class.forName()` never returns null.
Further I assume there are no class loading tricks where a different `TestEngine` class is a super class of this bundle's one? Therefore an equals (maybe even identity?) check should be sufficient, shouldn't it?
```suggestion
				return thisTestEngine.equals(bundleTestEngine);
```

Reply:
Yes, an identity comparison is sufficient here (fixed with 19e3dd5). I had two false assumptions:
* I accidentally expected `isAssignableFrom()` to return true for two instances of the same class loaded by different class loaders. Since that's the operation used by the `ServiceLoader` that fails when incompatible engine versions are used, I reused it here.
* I expected `Class.forName()` to return null if the engine is not on the classpath, but it will of course throw an exception. A scenario where I would expect this happens to is as follows: execute the runner without any JUnit 5 dependencies, but use the argument `-runasjunit5` and have some bundle specifying a service provider for "org.junit.platform.engine.TestEngine".

Reply:
> A scenario where I would expect this happens to is as follows: execute the runner without any JUnit 5 dependencies, but use the argument -runasjunit5 and have some bundle specifying a service provider for \"org.junit.platform.engine.TestEngine\".

I assume each bundle providing a TestEngine requires that class to implement it. And therefore, if the OSGi metadata are correct, it should depend on the providing bundle/package.
So I don't expect that to happen, but we'll see what specialties occur. ^^

Reply:
> So I don't expect that to happen

Me neither 🙂 I should have mentioned that the situation may only occur with invalid metadata (i.e., the declaration of a service that is actually not provided). And even for such a very unlikely, accidental case it might be good to fail in a more "useful" way than throwing an NPE.

And, of course, thanks for merging!